src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php

```diff
@@ -187,6 +187,31 @@ protected function performJoin($query = null)
         return $this;
     }
 
+    /**
+     * Resolves table name from a given string.
+     *
+     * @param  string  $class
+     * @return string
+     */
+    protected function resolveTableName($class)
+    {
+        if (! class_exists($class)) {
+            return $class;
+        }
+
+        $object = new $class;
+
+        if ($object instanceof Model) {
+            if ($object instanceof Pivot) {
+                $this->using($class);
+            }
+
+            return $object->getTable();
+        }
+
+        return $class;
```

Comment:
What's a scenario where this line is reached?

Reply:
Line (``212``) is reached when a given table name can't be resolved as an instance of a Model class, so in all other cases it just fallback to the old behavior.

Reply:
So for the rare case that a table name coincidentally is also a valid class name?

Reply:
My code checks if the given table name can be called as a class **and also extends Model / Pivot class**. The only rare case I could think of could be when user has a Pivot class **without a namespace** and uses "customers_profiles" as a second parameter for belongsToMany:

```php

// note: no namespace used

class customers_profiles extends Pivot
{
    protected $table = 'customers_profiles';
}
```

In this case, my code will see that's the valid Pivot class, will try to load Pivot model and take table name through getTable() method.

```php
dd((new customers_profiles)->getTable()); 
// gives back "customers_profiles" string, 
// or 
// auto-generated value "customers_profile" when user has not defined $table attribute
```

All other cases I could think of should fallback to an old behavior.

Reply:
p.s. I have slightly adjusted the code flow / naming.